File: core/core-api/src/main/java/com/coffee/gu/stamp/StampRevertManager.java
package com.coffee.gu.stamp;

import com.coffee.gu.CoreException;
import com.coffee.gu.ErrorType;
import com.coffee.gu.coupon.IssuedCoupon;
import com.coffee.gu.coupon.IssuedCouponRepository;
import com.coffee.gu.order.Order;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
public class StampRevertManager {

    private final StampRevertPlanner stampRevertPlanner;
    private final StampRepository stampRepository;
    private final IssuedCouponRepository issuedCouponRepository;
    private final StampRecorder stampRecorder;

    public StampRevertManager(StampRevertPlanner stampRevertPlanner, StampRepository stampRepository, IssuedCouponRepository issuedCouponRepository, StampRecorder stampRecorder) {
        this.stampRevertPlanner = stampRevertPlanner;
        this.stampRepository = stampRepository;
        this.issuedCouponRepository = issuedCouponRepository;
        this.stampRecorder = stampRecorder;

    }

    public void validateRevertable(Order order) {
        StampRevertPlan plan = stampRevertPlanner.plan(order);
        validate(plan);
    }

    @Transactional
    public void revert(Order order) {
        StampRevertPlan plan = stampRevertPlanner.plan(order);
        if (plan.isEmpty()) return;
        validate(plan);
        cancelRewardCoupons(plan);
        cancelStamps(plan);
        stampRecorder.recordCancel(order, plan);
    }

    private void validate(StampRevertPlan plan) {
        if (plan.hasUsedRewardCoupon()) {
            throw new CoreException(
                    ErrorType.INVALID_REQUEST,
                    "이미 사용된 리워드 쿠폰이 있어 주문을 취소할 수 없습니다."
            );
        }
    }

    private void cancelRewardCoupons(StampRevertPlan plan) {
        if (!plan.hasRewardCouponsToCancel()) return;
        plan.getRewardCouponsToCancel().forEach(IssuedCoupon::cancel);
        issuedCouponRepository.saveAll(plan.getRewardCouponsToCancel());
    }

    private void cancelStamps(StampRevertPlan plan) {
        plan.getStampsToCancel().forEach(Stamp::cancel);
        stampRepository.saveAll(plan.getStampsToCancel());
    }


}

File: core/core-api/src/main/kotlin/com/coffee/gu/stamp/StampRevertManager.kt
package com.coffee.gu.stamp

import com.coffee.gu.CoreException
import com.coffee.gu.ErrorType
import com.coffee.gu.coupon.IssuedCoupon
import com.coffee.gu.coupon.IssuedCouponRepository
import com.coffee.gu.order.Order
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class StampRevertManager(
    private val stampRevertPlanner: StampRevertPlanner,
    private val stampRepository: StampRepository,
    private val issuedCouponRepository: IssuedCouponRepository,
    private val stampRecorder: StampRecorder,
) {
    fun validateRevertable(order: Order) {
        val plan = stampRevertPlanner.plan(order)
        validate(plan)
    }

    @Transactional
    fun revert(order: Order) {
        val plan = stampRevertPlanner.plan(order)
        if (plan.isEmpty) return
        validate(plan)
        cancelRewardCoupons(plan)
        cancelStamps(plan)
        stampRecorder.recordCancel(order, plan)
    }

    private fun validate(plan: StampRevertPlan) {
        if (plan.hasUsedRewardCoupon()) {
            throw CoreException(
                ErrorType.INVALID_REQUEST,
                "이미 사용된 리워드 쿠폰이 있어 주문을 취소할 수 없습니다."
            )
        }
    }

    private fun cancelRewardCoupons(plan: StampRevertPlan) {
        if (!plan.hasRewardCouponsToCancel()) return
        plan.rewardCouponsToCancel.forEach { it.cancel() }
        issuedCouponRepository.saveAll(plan.rewardCouponsToCancel)
    }

    private fun cancelStamps(plan: StampRevertPlan) {
        plan.stampsToCancel.forEach { it.cancel() }
        stampRepository.saveAll(plan.stampsToCancel)
    }
}
